### PR TITLE
First basic scenarios

### LINKED ISSUE #103: Missing components
- heat pump, domestic
- heat pump, large-scale
- boiler, small

I have noticed that these three components are missing. Maybe you could add them in #91 @jnnr  as you will have to change the .yml files in that PR anyways..?

FYI I have added h2 import and export (shortage and excess) and ch4 import in #107   (update: added in 107)

### PR DESCRIPTION
**Pull request by jnnr**
Fix #103 
This PR introduces 3 basic scenarios. Dummy data will be replaced by real data. To test get the most recent data `raw/scalars_<base/low/high>_PR#91` and `raw/timeseries/ts_efficiency_heatpump_small.csv` and `raw/timeseries/ts_load-ch4.csv`, run `snakemake -j1 results/base-2050/postprocessed`

TODO
* components
  * [x] electricity-heatpump_small (heat pump, domestic)
  * [x] electricity-heatpump_large (heat pump, large-scale)
  * [x] ch4-boiler_small
  * [x] electricity-transmission
  * [x] electricity
    * [x] demand
    * [x] import
    * [x] export
    * [x] shortage
    * [x] cutrailment
  * [x] h2
    * [x] demand
    * [x] import
    * [x] export
    * [x] shortage
    * [x] excess
  * [x] ch4
    * [x] demand
    * [x] import
    * [x] export
    * [x] shortage
    * [x] excess
  * [x] <heat_central,heat_decentral>-<demand,shortage,excess>
  - [x] h2-cavern

Electricity export/import and h2/ch4 shortage is added in #123.

* timeseries
  * [x] renewables feedin
  * [x] electricity demand
  * [x] heat demand
  * [x] electricity bev demand

### CODE REVIEW

**Comment by SabineHaas** [comment]
@jnnr there's a failing test and a merge conflict. Would you like to solve this, before my review? 

The merge conflict might come from #107  let me know if you need help.

**Comment by jnnr** [comment]
@SabineHaas: I added a new scenario in 4ff0904 to test how the current scenario building process works. My impression: It works, but may become a bit hard to read with many scenario changes.

**Review by SabineHaas — COMMENTED** [review]
Thank you @jnnr for preparing these scenarios :) . I already have some comments and questions, but haven't finished to go through the whole PR, yet.

- You could add vehicle time series already (however, Brandenburg2019 is still a dummy).
- Some rules like `prepare_feedin` have to be run separately to get the `_resources`. This does not necessarily have to be fixed in this PR.
- How do I get `ts_load_heat.csv` as resource?  --> Maybe you can provide your `_resources` to me  - I guess that would be easiest.

FYI I needed to adapt `scalars_base_PR#91.csv` to be able to prepare the scalars with the current code: 
- replace comma with dot in one `var_value`
- use comma as separator
- remove columns source and comment (due to commas)

File: scenarios/base-2050.yml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+name: base-2050
+
+regions:
+  - BB
+  - B
+
+links:
+  - BB-B
+
```
Comment (SabineHaas) [inline scenarios/base-2050.yml:14]: Seems like this file is missing 'datetimeindex'. 

```
Traceback (most recent call last):
  File "scripts/build_datapackage.py", line 238, in <module>
    start=scenario_specs["datetimeindex"]["start"],
KeyError: 'datetimeindex'
```

scenario_specs is the loaded data of `scenarios/base-2050.yml`

**Review by SabineHaas — CHANGES_REQUESTED** [review]
Nice! Thank you @jnnr preparing these scenarios and also improving the code in this PR!

I've noticed that the rules that prepare resources (e.g. demands) are run automatically now when running `snakemake -j1 results/base-2050/postprocessed` :+1: (:

I was missing ts_load_h2 and just used the same file as ts_load_ch4 as they are both konstant. I have added it to /raw.

After the issue with the datetimeindex is fixed, I can test the full run.

File: oemof_b3/model/foreign_keys_update.yml [hunk 2]
```diff
@@ -1,35 +1,46 @@
...
 
 'from_to_bus': [
+    'ch4-boiler_large',
+    'ch4-boiler_small',
     'electricity-electrolyzer',
     'h2-gt',
     'electricity-heatpump_large',
     'electricity-heatpump_small',
     'electricity-pth',
+    'electricity-transmission',
```
[the remark this answers is not in the document]
Comment (jnnr) [reply oemof_b3/model/foreign_keys_update.yml:43]: Yes, exactly.

**Comment by SabineHaas** [comment]
@jnnr thanks for adding the datetimeindex to the scenarios, I have tested a full run now. 

I noticed that the results are quite unrealistic, as each demand is covered by only import/shortage. This is due to low costs for import and shortage. 
You could set high shortage costs to receive more realistic results. 
You could also add emission_factors as in `base-scenario.csv` given.  But I also understand if you don't want to make it more complicated. However, I think it would be convenient as this would show that all current functionalities/data works with the current code.

File: oemof_b3/tools/data_processing.py [hunk 3]
```diff
@@ -282,6 +295,74 @@ def aggregate_scalars(df, columns_to_aggregate, agg_method=None):
...
+    df_b_index_name = _df_b.index.name
+    df_b_columns = list(_df_b.columns)
+    if indicator:
+        df_b_columns.append("_merge")
+
+    # Warn if data is lost because of merge
+    a_not_b = pd.Index(_df_a.loc[:, on]).difference(pd.Index(_df_b.loc[:, on]))
+    print(
+        f"There are {len(a_not_b)} elements are in df_a but not in df_b and will be lost: {a_not_b}"
+    )
```
[the remark this answers is not in the document]
Comment (jnnr) [reply oemof_b3/tools/data_processing.py:335]: You are right. It should only appear if how == 'left', not if how == 'outer'. Will do that

File: oemof_b3/tools/data_processing.py [hunk 4]
```diff
@@ -282,6 +295,74 @@ def aggregate_scalars(df, columns_to_aggregate, agg_method=None):
...
+        _df_a,
+        on=on,
+        how=how,
+        indicator=indicator,
+        sort=False,
+    )
+
+    merged.index.name = df_b_index_name
+
+    # Where df_a contains no new data, use df_b
```
[the remark this answers is not in the document]
Comment (jnnr) [reply oemof_b3/tools/data_processing.py:351]: Yes.

File: scripts/create_empty_scalars.py [hunk 5]
```diff
@@ -139,4 +131,6 @@ def expand_scalars(df, column, where, expand):
             expand=["storage_capacity_cost_overnight", "storage_fixom_cost"],
         )
 
-    empty_scalars.to_csv(destination)
+    empty_scalars = sort_values(empty_scalars)
+
+    empty_scalars.to_csv(destination, sep=";")
```
[the remark this answers is not in the document]
Comment (jnnr) [reply scripts/create_empty_scalars.py:136]: You are right!

**Review by SabineHaas — APPROVED** [review]
Thanks for adding bev time series :) 
I have adapted the scalars and uploaded them to /raw: added emission_factors, emission_limit and adapted shortage costs to a very high value and exports costs to a low negative value.

I tested the scenario base-2050 but not the other scenarios.

**Comment by jnnr** [comment]
I tested base-2050 and base-2050-high_capacity_cost

**Pull request merged**